File: java_src/src/test/java/com/google/crypto/tink/internal/KeySerializerTest.java
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.internal;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import com.google.crypto.tink.InsecureSecretKeyAccess;
import com.google.crypto.tink.Key;
import com.google.crypto.tink.Parameters;
import com.google.crypto.tink.SecretKeyAccess;
import com.google.crypto.tink.util.Bytes;
import com.google.errorprone.annotations.Immutable;
import java.security.GeneralSecurityException;
import javax.annotation.Nullable;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Tests for {@link KeySerializer}. */
@RunWith(JUnit4.class)
public final class KeySerializerTest {

  @Immutable
  private static class ExampleKey extends Key {
    @Override
    public boolean equalsKey(Key k) {
      return k == this;
    }

    @Override
    @Nullable
    public Integer getIdRequirementOrNull() {
      return null;
    }

    @Override
    public Parameters getParameters() {
      throw new UnsupportedOperationException("Not needed in test");
    }
  }

  @Immutable
  private static class ExampleSerialization implements Serialization {
    @Override
    public Bytes getObjectIdentifier() {
      return Bytes.copyFrom(new byte[0]);
    }
  }

  private static ExampleSerialization serialize(ExampleKey k, @Nullable SecretKeyAccess access)
      throws GeneralSecurityException {
    SecretKeyAccess.requireAccess(access);
    return new ExampleSerialization();
  }

  @Test
  public void createSerializer_works() throws Exception {
    KeySerializer.create(
        KeySerializerTest::serialize, ExampleKey.class, ExampleSerialization.class);
  }

  @Test
  public void createSerializer_serializeKey_works() throws Exception {
    KeySerializer<ExampleKey, ExampleSerialization> serializer =
        KeySerializer.create(
            KeySerializerTest::serialize, ExampleKey.class, ExampleSerialization.class);
    assertThat(serializer.serializeKey(new ExampleKey(), InsecureSecretKeyAccess.get()))
        .isNotNull();
    assertThrows(
        GeneralSecurityException.class,
        () -> serializer.serializeKey(new ExampleKey(), /* access = */ null));
  }

  @Test
  public void createSerializer_classes_work() throws Exception {
    KeySerializer<ExampleKey, ExampleSerialization> serializer =
        KeySerializer.create(
            KeySerializerTest::serialize, ExampleKey.class, ExampleSerialization.class);
    assertThat(serializer.getKeyClass()).isEqualTo(ExampleKey.class);
    assertThat(serializer.getSerializationClass()).isEqualTo(ExampleSerialization.class);
  }
}

File: cc/internal/key_serializer_test.cc
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/internal/key_serializer.h"

#include <memory>
#include <string_view>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/memory/memory.h"
#include "absl/status/status.h"
#include "absl/types/optional.h"
#include "tink/insecure_secret_key_access.h"
#include "tink/internal/serialization.h"
#include "tink/internal/serialization_test_util.h"
#include "tink/internal/serializer_index.h"
#include "tink/key.h"
#include "tink/secret_key_access_token.h"
#include "tink/util/statusor.h"
#include "tink/util/test_matchers.h"

namespace crypto {
namespace tink {
namespace internal {
namespace {

using ::crypto::tink::test::IsOk;
using ::crypto::tink::test::StatusIs;
using ::testing::Eq;

TEST(KeySerializerTest, Create) {
  std::unique_ptr<KeySerializer> serializer =
      absl::make_unique<KeySerializerImpl<NoIdKey, NoIdSerialization>>(
          SerializeNoIdKey);

  EXPECT_THAT(serializer->Index(),
              Eq(SerializerIndex::Create<NoIdKey, NoIdSerialization>()));
}

TEST(KeySerializerTest, SerializeKey) {
  std::unique_ptr<KeySerializer> serializer =
      absl::make_unique<KeySerializerImpl<NoIdKey, NoIdSerialization>>(
          SerializeNoIdKey);

  NoIdKey key;
  util::StatusOr<std::unique_ptr<Serialization>> serialization =
      serializer->SerializeKey(key, InsecureSecretKeyAccess::Get());
  ASSERT_THAT(serialization, IsOk());
  EXPECT_THAT((*serialization)->ObjectIdentifier(), Eq(kNoIdTypeUrl));
}

TEST(KeySerializerTest, SerializePublicKeyNoAccessToken) {
  std::unique_ptr<KeySerializer> serializer =
      absl::make_unique<KeySerializerImpl<NoIdKey, NoIdSerialization>>(
          SerializeNoIdKey);

  NoIdKey public_key;
  util::StatusOr<std::unique_ptr<Serialization>> serialization =
      serializer->SerializeKey(public_key, absl::nullopt);
  ASSERT_THAT(serialization, IsOk());
  EXPECT_THAT((*serialization)->ObjectIdentifier(), Eq(kNoIdTypeUrl));
}

TEST(KeySerializerTest, SerializeKeyWithInvalidKeyType) {
  std::unique_ptr<KeySerializer> serializer =
      absl::make_unique<KeySerializerImpl<NoIdKey, NoIdSerialization>>(
          SerializeNoIdKey);

  IdKey key(/*id=*/123);
  util::StatusOr<std::unique_ptr<Serialization>> serialization =
      serializer->SerializeKey(key, InsecureSecretKeyAccess::Get());
  ASSERT_THAT(serialization.status(),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

}  // namespace
}  // namespace internal
}  // namespace tink
}  // namespace crypto
